csharp/src/Google.Protobuf.Test/JsonFormatterTest.cs

```diff
@@ -630,6 +630,18 @@ public void Wrappers_Standalone(System.Type wrapperType, object value, string ex
             Assert.AreEqual(expectedJson, JsonFormatter.Default.Format(populated));
         }
 
+        // See See https://github.com/protocolbuffers/protobuf/issues/11987
+        [Test]
+        public void JsonNamePriority()
+        {
+            // This tests both the formatter and the parser, but the issue was when parsing.
+            var original = new Issue11987Message { A = 10, B = 20, C = 30 };
+            var json = JsonFormatter.Default.Format(original);
+            AssertJson("{ 'b': 10, 'a': 20, 'd': 30 }", json);
+            var parsed = Issue11987Message.Parser.ParseJson(json);
```

Comment:
I'm curios, what happens if you try to parse "{ 'b': 10, 'a': 20, 'd': 30, 'c': 40 }"? I'm just wondering if on the map, there should only be a single entry per field with the key being JsonName if any, else Name. Or basically just have an internal EffectiveJsonName in the field descriptor?

Reply:
Good question - will check on Monday

Reply:
Right... our draft specification says that should be rejected, but currently the later field in the JSON will overwrite the earlier field. (For fields of a message type, it'll actually merge the two values.)

So the C# parser is in strict violation of the spec in that sense... as it happens, we *need* that non-compliant behavior for one CloudEvent parsing scenario at least. (We'll need to work out what to do about either having a "strict" parser setting or change the default in a new major version at some point.)

But I think the behavior with this PR is strictly better than before :)

Reply:
Yep, that all makes sense. Thanks!